### PR TITLE
Various fixes and improvements

### PR DESCRIPTION
**Pull request by npmccallum**
This PR contains a variety of improvements.

### CODE REVIEW

**Comment by npmccallum** [comment]
Please review this, but don't merge it. The `hw` tests are currently silently skipping the actual tests. I need to fix the runner (tomorrow).

File: src/types/sig.rs [hunk 1]
```diff
@@ -390,21 +390,11 @@ mod crypto {
...
-        }
-
+        let mut buf = [0u8; std::mem::size_of::<Signature>()];
         let mut file = File::open(path).unwrap();
-        file.read_exact(buf).unwrap();
+        file.read_exact(&mut buf).unwrap();
 
-        sig
+        unsafe { std::mem::transmute(buf) }
     }
```
[the remark this answers is not in the document]
Comment (npmccallum) [reply src/types/sig.rs:398]: @haraldh I'm going to drop that commit from this series. Please submit your fix as a separate PR with a proper signoff.

**Review by haraldh — DISMISSED** [review]

File: src/types/sig.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply src/types/sig.rs]: Never mind. I made my own version of it.

**Review by haraldh — DISMISSED** [review]

File: src/enclave/mod.rs [hunk 2]
```diff
@@ -15,3 +15,100 @@ mod ioctls;
...
+    use crate::types::page;
+    use crate::types::tcs::Tcs;
+
+    use std::num::NonZeroU32;
+
+    use lset::Span;
+    use primordial::Page;
+
+    #[allow(clippy::erasing_op)]
+    const CODE_OFFSET: usize = Page::size() * 0;
```
Comment (wgwoods) [inline src/enclave/mod.rs:32]: Comments for what these values mean would be v helpful

File: src/enclave/mod.rs [hunk 3]
```diff
@@ -15,3 +15,100 @@ mod ioctls;
...
+    const TCS_OFFSET: usize = Page::size() * 1;
+    const SSA_OFFSET: usize = Page::size() * 2;
+    const SSA_COUNT: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(1) };
+
+    /// Don't call this function. It has a non-standard ABI.
+    #[naked]
+    unsafe extern "C" fn code() -> ! {
+        asm!(
+            "0:",
```
Comment (wgwoods) [inline src/enclave/mod.rs:41]: comment explaining labels (here and the `.fill`)

File: src/enclave/mod.rs [hunk 4]
```diff
@@ -15,3 +15,100 @@ mod ioctls;
...
+        let span = Span {
+            start: 1 * 1024 * 1024 * 1024,
+            count: 1 * 1024 * 1024,
+        };
+
+        let mut builder = Builder::new(span).unwrap();
+
+        let segs: [Segment; 3] = [
+            Segment {
+                src: vec![unsafe { *(code as *const Page) }],
```
Comment (wgwoods) [inline src/enclave/mod.rs:68]: There should definitely be a comment explaining why you're doing this. Also maybe an apology.

File: src/enclave/mod.rs [hunk 5]
```diff
@@ -15,3 +15,100 @@ mod ioctls;
...
+                si: page::SecInfo::reg(page::Flags::R | page::Flags::W),
+            },
+        ];
+        builder.load(&segs).unwrap();
+
+        // Build the enclave.
+        let enclave = builder.build().unwrap();
+        let mut thread = Thread::new(enclave).unwrap();
+
+        let mut registers = enclave::Registers {
```
Comment (wgwoods) [inline src/enclave/mod.rs:93]: For fun, give different values to each register to ensure they're being incremented and not just set to 8

File: build.rs [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 fn main() {
-    if std::path::Path::new("/dev/sgx/enclave").exists() {
+    if std::path::Path::new("/dev/sgx_enclave").exists() {
```
Comment (wgwoods) [inline build.rs:4]: Should we maybe have `const SGX_ENCLAVE_DEVICE_PATH = ...;` with a docstring explaining why the path is what it is, how it gets populated, and how/whether it could be created dynamically? (This would be the place for the "hey your distro is broken, go fix your distro" comment")

File: src/types/sig.rs [hunk 7]
```diff
@@ -390,21 +390,15 @@ mod crypto {
...
-        let buf: &mut [u8];
-
-        unsafe {
-            sig = std::mem::MaybeUninit::uninit().assume_init();
-            buf = std::slice::from_raw_parts_mut(
-                &mut sig as *mut _ as *mut u8,
-                std::mem::size_of_val(&sig),
-            );
-        }
+        let mut sig = std::mem::MaybeUninit::<Signature>::uninit();
```
Comment (wgwoods) [inline src/types/sig.rs:393]: Add a comment explaining a) why this is safe ("..because we don't use it before writing..") and b) why the original code was screwing up ("don't do init after doing uninit" or whatever it was...)

File: src/enclave/mod.rs (hunk 3 above)
Comment (haraldh) [inline src/enclave/mod.rs:41]: Oh, btw, you should not use `0` or `1`. See https://doc.rust-lang.org/nightly/unstable-book/library-features/asm.html#labels

> Moreover, due to an llvm bug, you shouldn't use labels exclusively made of 0 and 1 digits, e.g. 0, 11 or 101010, as they may end up being interpreted as binary values.

**Comment by wgwoods** [comment]
Everything here makes sense to me - overall I'd make two general suggestions:

1. any `unsafe { ... }` block should have a `// SAFETY:` note accompanying it, explaining to some future auditor why the code is actually safe (i.e. what behavior would _actually_ be unsafe, and how do we know this code isn't doing that)
2. any code that's a temporary workaround for some other problem should have a `// TODO:` or `// WORKAROUND:` comment that references the issue being worked around (so we can find & remove stale workarounds later)

File: build.rs (hunk 6 above)
Comment (npmccallum) [reply build.rs:4]: It looks like doing this will cause more pain than it solves because we have references to this path across compilation units with no shared code. Creating a shared crate for this problem is not worth the overhead.

**Comment by npmccallum** [comment]
> tl;dr: this all makes sense - add some comments and I'll approve
> 
> General comments:
> 
> 1. any `unsafe { ... }` block should have a `// SAFETY:` note accompanying it, explaining to some future auditor why the code is actually safe (i.e. what behavior would _actually_ be unsafe, and how do we know this code isn't doing that)

I love this as a policy. I'd love it even more as a clippy lint. Clippy already enforces `/// # Safety` doc strings on public unsafe functions.

> 2. any code that's a temporary workaround for some other problem should have a `// TODO:` or `// WORKAROUND:` comment that references the issue being worked around (so we can find & remove stale workarounds later)

+1 for `TODO` since lots of tooling can do things with it.

> 3. the end-to-end test in `enclave/mod.rs` is illustrative but nontrivial to follow, and could really use some inline explanatory comments for what it's doing - I'd like to know which parts are needed because of SGX requirements and which things are workarounds/hacks (`vec![unsafe { *(code as *const Page) }]` definitely deserves a few words...)

Fixed.

**Review by wgwoods — APPROVED** [review]
Looks good!

**Review by haraldh — APPROVED** [review]
